src/nanotron/data/dataset_builder.py

```diff
@@ -0,0 +1,214 @@
+
+    def build(self) -> List[Union[BlendedNanoset, Nanoset]]:
+        """Build all dataset splits according to the provided data_path(s)
+
+        This method is distributed-aware and must be called on all ranks.
+
+        The dataset splits returned can vary according to the config. Supply config.data_path and
+        config.split to build BlendedNanoset and/or Nanoset splits from the same
+        distribution.
+
+        Returns:
+            List[Union[BlendedNanoset, Nanoset]]: A list of either Nanoset or BlendedNanoset per split
+        """
+
+        data_path = self.config.data_path
+        split = self.config.split_vector
+
+        # Single Nanoset
+        if isinstance(data_path, str):
+            return self.build_nanoset_dataset_splits(data_path, split, self.config.split_num_samples)
+
+        # Blended Nanoset
+        prefix_per_dataset = list(data_path.keys())
+        weight_per_dataset = normalize(list(data_path.values()))
+
+        # NOTE: Use 0.5% target margin to ensure that that we do not exhaust the indices of any dataset in the Blend
+        sizes_per_dataset = [
+            [
+                int(math.ceil(target_num_samples * weight * 1.005))
+                for target_num_samples in self.config.split_num_samples
+            ]
+            for weight in weight_per_dataset
+        ]
+
+        nanoset_datasets = [[] for _ in range(len(Split))]
+
+        for i in range(len(prefix_per_dataset)):
+            nanoset_datasets_split = self.build_nanoset_dataset_splits(
+                prefix_per_dataset[i], split, sizes_per_dataset[i]
+            )
+
+            for j in range(len(nanoset_datasets_split)):
+                nanoset_datasets[j].append(nanoset_datasets_split[j])
+
+        # Sum over all contributing datasets, per split
+        size_per_split = list(map(sum, zip(*sizes_per_dataset)))
+
+        blended_nanosets = []
+
+        for i in range(len(nanoset_datasets)):
+            is_none = (_ is None for _ in nanoset_datasets[i])
+
+            if split[i] == 0.0:
+                assert all(is_none)
+                blended_nanosets.append(None)
+            else:
+                assert not any(is_none)
+                blended_nanosets.append(
+                    self.build_generic_dataset(
+                        BlendedNanoset,
+                        nanoset_datasets[i],
+                        weight_per_dataset,
+                        size_per_split[i],
+                        self.config,
+                    )
+                )
+
+        return blended_nanosets
+
+    def build_nanoset_dataset_splits(
+        self,
+        path_prefix: str,
+        split: List[float],
+        sizes: List[int],
+    ) -> List[Nanoset]:
+        """Build each Nanoset split from a single MMapIndexedDataset
+
+        Args:
+            path_prefix (str): The MMapIndexedDataset .bin and .idx file prefix
+
+            split (List[float]): The dataset split ratios (must sum to 1.00)
```

Comment:
use `weights` as the argument name?

Reply:
and assert it sum up to 1

Reply:
There are two different distributions:
1. The train, valid, and test split. This is the one you are mentioning, which is entered in the split field of the .yaml file. When creating the `NanosetConfig`, we already parse and normalize this string.

2. The weights of each dataset for the `BlendedNanoset`. These are specified in the .yaml file, where the values are the weights. In the `dataset_builder`, we use them to compute the number of samples for each `Nanoset`.

Reply:
Should we add the assert considering that we always normalize it?

Reply:
> Should we add the assert considering that we always normalize it?
Yes, it would be nice from an end-user perspective, so they don't have unexpected normalization (like the weights they get should be the weights that they enter)

Reply:
[Done!](https://github.com/TJ-Solergibert/nanotron/blob/f163c1d6574b5f8f1d7e9d51caa9f006ec35660a/src/nanotron/data/dataset_builder.py#L115)